File: test/js/personalization_manufacture.test.mjs
import test from "node:test";
import assert from "node:assert/strict";

import {
  manufactureTraceSchema,
  personalizationContextSchema,
} from "../../priv/static/ash_surface_runtime.mjs";

function validContext() {
  return {
    contextId: "pc_demo",
    exactSubject: "person:demo",
    facets: [
      {
        facetId: "pf_goal",
        dimension: "life:outcome",
        valueRef: "life:outcome:consistency",
        source: "USER_STATED",
        standing: "ALIVE",
        falsifier: null,
        evidenceRefs: ["evidence:user-selection"],
      },
    ],
    consentRef: "consent:subject-only",
    evidenceRefs: ["evidence:user-selection"],
    standing: "ALIVE",
    privacyScope: "SUBJECT_PRIVATE",
    shareScope: "SUBJECT_ONLY",
    stateDigest: "digest-profile",
    authorityBoundary: "OBSERVE",
    doAuthority: false,
  };
}

function validTrace() {
  return {
    traceId: "mt_demo",
    exactSubject: "person:demo",
    artifactRef: "practice:devotional:perseverance",
    manufacturerIdentity: "zoe:personalization:semantic-map",
    humanSummary: "Built from the admitted goal and devotional semantics.",
    observedRefs: ["o:goal"],
    admittedRefs: ["o:goal"],
    groundedRefs: ["o:goal"],
    boundedRefs: ["o:goal"],
    alignedRefs: ["o:goal"],
    oStarRefs: ["o:goal"],
    receiptRefs: ["receipt:manufacture:001"],
    falsifiers: ["goal admission is withdrawn"],
    standing: "ALIVE",
    stateDigest: "digest-trace",
    equation: "A=mu(O*)",
    authorityBoundary: "OBSERVE",
    doAuthority: false,
  };
}

test("subject-private USER_STATED personalization context is admitted", () => {
  const context = personalizationContextSchema.parse(validContext());
  assert.equal(context.facets[0].source, "USER_STATED");
  assert.equal(context.privacyScope, "SUBJECT_PRIVATE");
  assert.equal(context.shareScope, "SUBJECT_ONLY");
  assert.equal(context.doAuthority, false);
});

test("INFERRED personalization cannot enter without a falsifier", () => {
  const context = validContext();
  context.facets[0].source = "INFERRED";
  context.facets[0].standing = "UNKNOWN";
  context.facets[0].falsifier = null;

  assert.equal(personalizationContextSchema.safeParse(context).success, false);
});

test("personalization context cannot acquire DO authority", () => {
  const context = validContext();
  context.doAuthority = true;

  assert.equal(personalizationContextSchema.safeParse(context).success, false);
});

test("receipted A=mu(O*) manufacture trace is admitted", () => {
  const trace = manufactureTraceSchema.parse(validTrace());

  assert.equal(trace.equation, "A=mu(O*)");
  assert.deepEqual(trace.oStarRefs, ["o:goal"]);
  assert.deepEqual(trace.receiptRefs, ["receipt:manufacture:001"]);
  assert.equal(trace.doAuthority, false);
});

test("O* must be inside observed/admitted/grounded/bounded/aligned intersection", () => {
  const trace = validTrace();
  trace.boundedRefs = [];

  assert.equal(manufactureTraceSchema.safeParse(trace).success, false);
});

test("ALIVE manufacture trace without receipt is refused", () => {
  const trace = validTrace();
  trace.receiptRefs = [];

  assert.equal(manufactureTraceSchema.safeParse(trace).success, false);
});

test("manufacture provenance cannot acquire authority", () => {
  const trace = validTrace();
  trace.doAuthority = true;

  assert.equal(manufactureTraceSchema.safeParse(trace).success, false);
});

File: test/ash_surface/personalization_manufacture_test.exs
defmodule AshSurface.PersonalizationManufactureTest do
  use ExUnit.Case, async: true

  alias AshSurface.{ManufactureTrace, PersonalizationContext}

  test "USER_STATED profile facet remains distinct from inference and subject-private" do
    context =
      PersonalizationContext.create(
        "person:demo",
        [
          %{
            dimension: "life:outcome",
            value_ref: "life:outcome:consistency",
            source: :USER_STATED,
            standing: :ALIVE,
            evidence_refs: ["evidence:user-selection"]
          }
        ],
        consent_ref: "consent:subject-only",
        standing: :ALIVE
      )

    map = PersonalizationContext.to_map(context)
    [facet] = map["facets"]

    assert facet["source"] == "USER_STATED"
    assert facet["standing"] == "ALIVE"
    assert map["privacyScope"] == "SUBJECT_PRIVATE"
    assert map["shareScope"] == "SUBJECT_ONLY"
    assert map["authorityBoundary"] == "OBSERVE"
    assert map["doAuthority"] == false
  end

  test "INFERRED profile facet requires an explicit falsifier" do
    assert_raise ArgumentError, ~r/requires a falsifier/, fn ->
      PersonalizationContext.create(
        "person:demo",
        [
          %{
            dimension: "life:outcome",
            value_ref: "life:outcome:consistency",
            source: :INFERRED,
            standing: :UNKNOWN
          }
        ]
      )
    end
  end

  test "personalization context digest is invariant to facet order" do
    left = %{
      dimension: "life:outcome",
      value_ref: "life:outcome:consistency",
      source: :USER_STATED,
      standing: :ALIVE
    }

    right = %{
      dimension: "practice:preference",
      value_ref: "practice:audio",
      source: :OBSERVED,
      standing: :PARTIAL_ALIVE
    }

    a = PersonalizationContext.create("person:demo", [left, right])
    b = PersonalizationContext.create("person:demo", [right, left])

    assert a.state_digest == b.state_digest
  end

  test "A=mu(O*) refuses an O* reference outside the full admitted intersection" do
    assert_raise ArgumentError, ~r/every O\* reference/, fn ->
      ManufactureTrace.create(
        "person:demo",
        "artifact:devotional-candidate",
        "manufacturer:semantic-map",
        observed_refs: ["o:goal"],
        admitted_refs: ["o:goal"],
        grounded_refs: ["o:goal"],
        bounded_refs: [],
        aligned_refs: ["o:goal"],
        o_star_refs: ["o:goal"]
      )
    end
  end

  test "ALIVE manufacture trace requires a receipt" do
    assert_raise ArgumentError, ~r/requires at least one receipt/, fn ->
      ManufactureTrace.create(
        "person:demo",
        "artifact:devotional-candidate",
        "manufacturer:semantic-map",
        observed_refs: ["o:goal"],
        admitted_refs: ["o:goal"],
        grounded_refs: ["o:goal"],
        bounded_refs: ["o:goal"],
        aligned_refs: ["o:goal"],
        o_star_refs: ["o:goal"],
        standing: :ALIVE
      )
    end
  end

  test "receipted A=mu(O*) trace exposes provenance without gaining authority" do
    trace =
      ManufactureTrace.create(
        "person:demo",
        "artifact:devotional-candidate",
        "manufacturer:semantic-map",
        observed_refs: ["o:goal"],
        admitted_refs: ["o:goal"],
        grounded_refs: ["o:goal"],
        bounded_refs: ["o:goal"],
        aligned_refs: ["o:goal"],
        o_star_refs: ["o:goal"],
        receipt_refs: ["receipt:manufacture:001"],
        falsifiers: ["goal admission is withdrawn"],
        human_summary: "Built from the admitted goal and devotional semantics.",
        standing: :ALIVE
      )

    map = ManufactureTrace.to_map(trace)

    assert map["equation"] == "A=mu(O*)"
    assert map["oStarRefs"] == ["o:goal"]
    assert map["receiptRefs"] == ["receipt:manufacture:001"]
    assert map["standing"] == "ALIVE"
    assert map["authorityBoundary"] == "OBSERVE"
    assert map["doAuthority"] == false
  end
end
